File: goods-review-frontend/src/main/scala/ru/goodsreview/frontend/core/HttpServerInitializer.scala
package ru.goodsreview.frontend.core

import org.springframework.beans.factory.InitializingBean
import org.mortbay.jetty.{Connector, Server, Handler}
import reflect.BeanProperty
import org.mortbay.thread.QueuedThreadPool
import org.mortbay.jetty.handler.HandlerCollection
import org.mortbay.jetty.nio.SelectChannelConnector
import org.apache.log4j.Logger

/**
 * @author Dmitry Batkovich
 */
class HttpServerInitializer extends InitializingBean {

  @BeanProperty
  var port: Int = _

  @BeanProperty
  var maxThreads: Int = _

  @BeanProperty
  var handlers: Array[Handler] = _

  private val server: Server = new Server()

  override def afterPropertiesSet() {
    val st = System.currentTimeMillis
    try {
      val connector = new SelectChannelConnector
      connector.setPort(port)

      val handlerCollection = new HandlerCollection()
      handlerCollection.setHandlers(handlers)

      server.setConnectors(Array(connector))
      server.setThreadPool(new QueuedThreadPool(maxThreads))
      server.setHandler(handlerCollection)
      server.start()

      log.info("Server started: " + (System.currentTimeMillis - st) + " ms")
    } catch {
      case e: Exception => log.error("Could not initialize server: ", e)
    }
  }

  def log = Helper.log
}

private object Helper {
  val log = Logger getLogger classOf[HttpServerInitializer]
}

File: goods-review-frontend/src/main/java/ru/goodsreview/frontend/core/HttpServerInitializer.java
package ru.goodsreview.frontend.core;

import org.mortbay.jetty.Connector;
import org.mortbay.jetty.Handler;
import org.mortbay.jetty.Server;
import org.mortbay.jetty.bio.SocketConnector;
import org.mortbay.jetty.nio.SelectChannelConnector;
import org.mortbay.jetty.servlet.Holder;
import org.springframework.beans.factory.DisposableBean;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Required;

import java.util.List;

/**
 * @author Artemii Chugreev achugr@yandex-team.ru
 *         07.10.12
 */
public class HttpServerInitializer implements InitializingBean, DisposableBean {

    private int port;

    private Handler[] handlers;

    @Required
    public void setPort(final int port) {
        this.port = port;
    }

    @Required
    public void setHandlers(final Handler[] handlers) {
        this.handlers = handlers;
    }

    private Server server;

    @Override
    public void afterPropertiesSet() throws Exception {
        server = new Server();
        Connector connector = new SelectChannelConnector();
        connector.setPort(port);
        server.addConnector(connector);
        server.setHandlers(handlers);
        server.start();
    }

    @Override
    public void destroy() throws Exception {
        server.stop();
    }
}
